### PR TITLE
Progress bar

### PR DESCRIPTION
**Pull request by ilastikdev**
if gui is present it displays a progress dialog box
if not it will display progress on command line

### CODE REVIEW

**Comment by chaubold** [comment]
Right now your `exportStep` and `exportProgress` default to nothing, and that clutters the code with `if not self.exportStep==None` (which should have been `if self.exportStep is not None` btw.).

For one, this should default to the command line progress bar as before. And I'm not really happy with the way the code looks. Perhaps it is because the variable names are a little confusing.

Do you know the [visitor design pattern](https://www.tutorialspoint.com/design_pattern/visitor_pattern.htm)? To make this nice from a software design perspective, you should call the `exportStep` things something like a `progressVisitor`. Then create a base class for the visitor that defines which methods it provides, make a default implementation (e.g. `CommandLineProgressVisitor`) to which this is set by default, and in the hypotheses graph provide a method to set a user defined visitor (e.g. `HypothesesGraph.setProgressVisitor`). And lastly on the ilastik side you can derive the `IlastikProgressVisitor` and pass that in as needed.

And to get rid of all the `is not None` checks throughout the code we should not allow the visitor to be set to `None` within `HypothesesGraph.setProgressVisitor`.

**Comment by ilastikdev** [comment]
visitor functions changed to visitor class
DefaultProgressVisitor and CommandLineProgressVisitor are defined in hytra
one can also setProgressVisitor from frontend

File: hytra/core/jsongraph.py [hunk 1]
```diff
@@ -241,8 +242,7 @@ def __init__(self,
                  model_filename=None, 
                  weights_filename=None, 
                  result_filename=None,
-                 exportStepFunction=None,
-                 exportProgressFunction=None):
+                 progressVisitor=ProgressVisitor()):
```
Comment (chaubold) [inline hytra/core/jsongraph.py]: This should be `DefaultProgressVisitor()`.

File: hytra/util/progressbar.py [hunk 2]
```diff
@@ -25,3 +25,60 @@ def show(self, increase=1):
...
+                sys.stdout.write('\n')
+                sys.stdout.flush()
+        except IOError:
+            pass
+
+class ProgressVisitor(DefaultProgressVisitor):
+    def __init__(self):
+        pass
+
+    def setProgressVisitor(self,progressVisitor):
```
Comment (chaubold) [inline hytra/util/progressbar.py]: I don't get the `ProgressVisitor`. It subclasses `DefaultProgressVisitor` which provides only empty methods, so setting `self.progressVisitor` here doesn't have any consequences, same as setting all the other members. I think that class should be removed. Also, it is not used from the ilastik side but there you did what one would expect -- you derived a special visitor for the Gui progress bar.

**Review by chaubold — CHANGES_REQUESTED** [review]

**Comment by ilastikdev** [comment]
removed... just a left over from the time I planed to use setProgressVisitor ... but that is actually not needed

**Review by chaubold — APPROVED** [review]

**Pull request merged**